### PR TITLE
Phoenix with LiveView (SVG clock) example

### PR DESCRIPTION
**Pull request by thbar**
This PR adds a `Mix.install/2`-based example of a Phoenix, LiveView-enabled (with esbuild) app as a little SVG clock.

I have kept some js and css as a demonstration of esbuild use.

Some extra cleaning and shortening can probably be done ^_^

https://user-images.githubusercontent.com/10141/131264446-ec431754-fe7b-4647-bada-099448a3b2b9.mp4

Many thanks to @wojtekmach, @teamon, @josevalim and @hauleth for their help and inspiration (see https://github.com/wojtekmach/mix_install_examples/issues/2#issuecomment-906343911).

### CODE REVIEW

File: phoenix_liveview/assets/app.css [hunk 1]
```diff
@@ -0,0 +1,15 @@
+html, body {
```
Comment (thbar) [inline phoenix_liveview/assets/app.css:1]: I have kept a bit of CSS to demonstrate the use of the `esbuild` pipeline. I could probably find a way to move it inside the "single file app", but it would generate some extra complexity which is probably not very relevant.

File: phoenix_liveview/assets/app.js [hunk 2]
```diff
@@ -0,0 +1,7 @@
+import "./app.css"
+import { Socket } from "phoenix"
```
Comment (thbar) [inline phoenix_liveview/assets/app.js:2]: The imports from hex packages has been a source of pain (see `NODE_PATH` hack further below), but it works.

File: phoenix_liveview/mix_install_deps_backport.exs [hunk 3]
```diff
@@ -0,0 +1,39 @@
+defmodule InstallFolderTemporaryBackport do
```
Comment (thbar) [inline phoenix_liveview/mix_install_deps_backport.exs]: This code is necessary to mimic the computation made by `Mix.install/2` to create the install folders, from where the app needs to find the `package.json` of each JS dependency.

Ideally a bit of refactoring of `Mix.install/2` itself would help expose the correct paths to the app, and remove the need for this.

File: phoenix_liveview/phoenix_liveview.exs [hunk 4]
```diff
@@ -0,0 +1,134 @@
+Application.put_env(:phoenix, :json_library, Jason)
```
Comment (thbar) [inline phoenix_liveview/phoenix_liveview.exs:1]: Required to avoid a warning

File: phoenix_liveview/phoenix_liveview.exs [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+Application.put_env(:esbuild, :version, "0.12.18")
```
Comment (thbar) [inline phoenix_liveview/phoenix_liveview.exs:2]: Similarly, this must be set before `Mix.install/2` or there will be a warning.

File: phoenix_liveview/phoenix_liveview.exs [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+    :jason,
+    :plug_cowboy,
+    {:phoenix_live_view, "~> 0.16.0"},
+    {:esbuild, "~> 0.2"}
+  ]
+)
+
+Code.require_file("mix_install_deps_backport.exs")
+
+Application.put_env(:esbuild, :default,
```
Comment (thbar) [inline phoenix_liveview/phoenix_liveview.exs:12]: This part has been a bit of pain ; one must not move this before `Mix.install/2`, or the defaults are actually erased, and the user gets cryptic error messages due to incorrect (incomplete) `esbuild` CLI call.

File: phoenix_liveview/phoenix_liveview.exs [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+  args: ~w(app.js --bundle --target=es2016 --outdir=../priv/static/assets),
+  cd: Path.expand("assets", __DIR__),
+  env: %{"NODE_PATH" => InstallFolderTemporaryBackport.determine_build_folder(deps) <> "/deps"}
```
Comment (thbar) [inline phoenix_liveview/phoenix_liveview.exs]: As noted before, this will be much improved if `Mix.install/2` exposes the computed `deps` to the script.

File: phoenix_liveview/phoenix_liveview.exs [hunk 8]
```diff
@@ -0,0 +1,134 @@
...
+
+    {:ok, put_date(socket)}
+  end
+
+  def handle_info(:tick, socket) do
+    {:noreply, put_date(socket)}
+  end
+
+  defp put_date(socket) do
+    assign(socket, date: NaiveDateTime.local_now())
```
Comment (thbar) [inline phoenix_liveview/phoenix_liveview.exs:74]: `NaiveDateTime.local_now()` does not currently support fractional seconds, otherwise I would have used that to make a smoother (sub-second) needle move.

File: phoenix_liveview/phoenix_liveview.exs [hunk 9]
```diff
@@ -0,0 +1,134 @@
...
+  end
+end
+
+defmodule LiveClock.LayoutView do
+  use Phoenix.View,
+    root: "templates",
```
Comment (thbar) [inline phoenix_liveview/phoenix_liveview.exs:80]: I have not yet found a way to set the template programmatically via a function, something which would have helped  remove the layout standalone file.

File: phoenix_liveview/phoenix_liveview.exs [hunk 10]
```diff
@@ -0,0 +1,134 @@
...
+
+  use Phoenix.HTML
+  alias LiveClock.Router.Helpers, as: Routes
+end
+
+defmodule LiveClock.Router do
+  use Phoenix.Router
+  import Phoenix.LiveView.Router
+
+  pipeline :browser do
```
Comment (thbar) [inline phoenix_liveview/phoenix_liveview.exs:112]: A part of what's below could probably be removed, but I have kept CSRF for now.

File: phoenix_liveview/phoenix_liveview.exs [hunk 11]
```diff
@@ -0,0 +1,134 @@
...
+  ]
+)
+
+defmodule LiveClock.ClockLive do
+  use Phoenix.LiveView
+  use Phoenix.HTML
+  import Phoenix.LiveView.Helpers
+
+  def render(assigns) do
+    ~L"""
```
Comment (thbar) [inline phoenix_liveview/phoenix_liveview.exs:35]: I had to use `~L` instead of `~H` because I could not get `~H` to work properly (it complained about unclosed attribute).

**Comment by josevalim** [comment]
Thanks @thbar! Since this is a demo of Mix.install, I think we should rather have all JavaScript+CSS inline and keep everything in a single file. I don't think esbuild is too important for this case. :)

File: phoenix_liveview/phoenix_liveview.exs (hunk 11 above)
Comment (wojtekmach) [reply phoenix_liveview/phoenix_liveview.exs:35]: fyi: https://github.com/phoenixframework/phoenix/issues/4449#issue-982872797

File: phoenix_liveview/phoenix_liveview.exs (hunk 4 above)
Comment (wojtekmach) [reply phoenix_liveview/phoenix_liveview.exs:1]: fwiw, we plan to address it with Mix.install for the next Elixir release. We envision this API:

```elixir
Mix.install(
  [
    :phoenix
  ],
  config: [
    phoenix: [json_library: Jason]
  ]
)
```

we're gonna use the config as part of the cache key, if it changes, the cache is stale so everything is recompiled.

File: phoenix_liveview/mix_install_deps_backport.exs (hunk 3 above)
Comment (wojtekmach) [reply phoenix_liveview/mix_install_deps_backport.exs]: Do you have some concrete changes to Mix.install in mind?

One idea is the following. Mix.install can already be configured via the `MIX_INSTALL_DIR`. Maybe after Mix installing, we'd set the env variable?

**Comment by thbar** [comment]
A quick update: I am fairly busy at the moment, but will comment back soon. Please keep the PR open if that's OK for you. Thanks!

File: phoenix_liveview/mix_install_deps_backport.exs (hunk 3 above)
Comment (thbar) [reply phoenix_liveview/mix_install_deps_backport.exs]: I will try to explain what is problematic with setting `MIX_INSTALL_DIR` only.

If I run `MIX_INSTALL_DIR=some-folder elixir phoenix_liveview.exs`, the install folder layout will be:

```
❯ tree some-folder -L 3
some-folder
└── elixir-1.12.2-erts-12.0.3
    └── 3592bd731a480b3e42d3e6048caaf789
        ├── _build
        ├── deps
        └── mix.lock
```

This means that while running `MIX_INSTALL_DIR=some-folder elixir phoenix_liveview.exs`,  the running program needs to figure out where is the `deps` folder, and so recompute the whole `elixir-1.12.2-erts-12.0.3/3592bd731a480b3e42d3e6048caaf789`.

Once the caller knows the folder, they can configure whatever they need to build the assets. Here if I hardcode what is necessary with the `tree` output above, it is:

```elixir
Application.put_env(:esbuild, :default,
  args: ~w(app.js --bundle --target=es2016 --outdir=../priv/static/assets),
  cd: Path.expand("assets", __DIR__),
  env: %{
    "NODE_PATH" =>
      Path.expand("some-folder/elixir-1.12.2-erts-12.0.3/3592bd731a480b3e42d3e6048caaf789/deps")
  }
```

To make it easy, it would be necessary to expose (instead of `MIX_INSTALL_DIR`) the `dir` variable to the application, as computed here in Mix:

https://github.com/elixir-lang/elixir/blob/3a7a598c255fdc1e719bc410f77071f7cdf9dd95/lib/mix/lib/mix.ex#L626-L628

(copy-pasted for ease of read:)

```elixir
  id = deps |> :erlang.term_to_binary() |> :erlang.md5() |> Base.encode16(case: :lower)
    version = "elixir-#{System.version()}-erts-#{:erlang.system_info(:version)}"
    dir = Path.join([installs_root, version, id])
```

So to summarise, it would be useful, for anyone using assets coming from `deps` (including phoenix & liveview), to programmatically expose the exact subfolder of installation.

@wojtekmach hope this is clearer, let me know if you think this would be doable!

**Comment by thbar** [comment]
> Since this is a demo of Mix.install, I think we should rather have all JavaScript+CSS inline and keep everything in a single file. I don't think esbuild is too important for this case. :)

@josevalim the `app.js` defined in the PR (https://github.com/wojtekmach/mix_install_examples/pull/3/files#diff-6590773b629f56a39d69eab83f7400a730b8da5d6497b495a4293932d64c75c8) uses stuff like `import { LiveSocket } from "phoenix_live_view"`.

I assumed that either `esbuild` or `webpack` would be necessary for this.

Do you mean it could not be the case? I will double-check anyway.

File: phoenix_liveview/mix_install_deps_backport.exs (hunk 3 above)
Comment (thbar) [reply phoenix_liveview/mix_install_deps_backport.exs]: This is now irrelevant, because I have applied @josevalim suggestion in 708e1de6b4710fb8ed55b9eefca71d01ba7cef36 (`Path.expand(Application.app_dir(:phoenix, "../../../../deps"))`) to avoid all this code.

File: phoenix_liveview/phoenix_liveview.exs (hunk 9 above)
Comment (thbar) [reply phoenix_liveview/phoenix_liveview.exs:80]: Handled via 6bb6a90

**Comment by wojtekmach** [comment]
@thbar sorry for radio silence. I'd really like to stick to a single file if possible. If not, let's hold off until it is possible. :) I just found https://twitter.com/TheGazler/status/1467900189906194433, maybe it helps for what you want to achieve?

**Comment by thbar** [comment]
@wojtekmach no problem! And no hurry :smile: ; definitely agree (I saw the tweet btw), if you are comfortable with leaving this open I will eventually come back to it (else I'll just reopen with a finished version as time permits).

**Comment by wojtekmach** [comment]
We can leave this open and there is of course no rush!

**Comment by wojtekmach** [comment]
Closing since we have #11!

**Comment by thbar** [comment]
Sure, no problem!

— Thibaut
